dojo/utils.py

```diff
@@ -2132,6 +2137,75 @@ def is_finding_groups_enabled():
     return get_system_setting("enable_finding_groups")
 
 
+class async_delete():
+    def __init__(self, *args, **kwargs):
+        self.mapping = {
+            'Product_Type': [
+                (Endpoint, 'product__prod_type'),
+                (Finding, 'test__engagement__product__prod_type'),
+                (Test, 'engagement__product__prod_type'),
+                (Engagement, 'product__prod_type'),
+                (Product, 'prod_type')],
+            'Product': [
+                (Endpoint, 'product'),
+                (Finding, 'test__engagement__product'),
+                (Test, 'engagement__product'),
+                (Engagement, 'product')],
+            'Engagement': [
+                (Finding, 'test__engagement'),
+                (Test, 'engagement')],
+            'Test': [(Finding, 'test')]
+        }
+
+    @dojo_async_task
+    @app.task
+    def delete_chunk(self, objects, **kwargs):
+        for object in objects:
+            object.delete()
+
+    @dojo_async_task
+    @app.task
+    def delete(self, object, **kwargs):
+        logger.debug('ASYNC_DELETE: Deleting ' + self.get_object_name(object) + ': ' + str(object))
+        model_list = self.mapping.get(self.get_object_name(object), None)
+        if model_list:
+            # The object to be deleted was found in the object list
+            self.crawl(object, model_list)
+        else:
+            # The object is not supported in async delete, delete normally
+            logger.debug('ASYNC_DELETE: ' + self.get_object_name(object) + ' async delete not supported. Deleteing normally: ' + str(object))
+            object.delete()
+
+    @dojo_async_task
+    @app.task
+    def crawl(self, object, model_list, **kwargs):
+        logger.debug('ASYNC_DELETE: Crawling ' + self.get_object_name(object) + ': ' + str(object))
+        for model_info in model_list:
+            model = model_info[0]
+            model_query = model_info[1]
+            filter_dict = {model_query: object}
+            objects_to_delete = model.objects.filter(**filter_dict)
+            logger.debug('ASYNC_DELETE: Deleting ' + str(len(objects_to_delete)) + ' ' + self.get_object_name(model) + 's in chunks')
+            chunks = self.chunk_list(model, objects_to_delete)
+            for chunk in chunks:
+                print('deleting', len(chunk), self.get_object_name(model))
+                self.delete_chunk(chunk)
```

Comment:
> Is there still a benefit to do chunked deletes? 

Absolutely! Since Django query sets are lazy loaded, they entire object is not actually loaded into memory until it is accessed. By chunking, only the chunked amount of objects are loaded into memory on each worker. So if only 8 workers are permitted to run at a time, at a chunk size of 100, that is only 800 objects loaded into memory at a time. In the experience that motivated this PR, the product had 38 thousand endpoints that were all loaded into memory at once. It crashed the uwsgi container.

> Deleting in parallel might be slightly faster

I am unsure how to do this without chunking. 

> but isn't there a risk of timing issues where it tries to start deleting Tests when the Findings in those tests aren't yet deleted?

This is mostly solved by the message broker. It handles messages (tasks) in a linear fashion so that all finding deleting tasks will be scheduled and ran before the test deleting tasks can be ran. This does not account for when the last finding deleting tasks does not end before the test deleting task starts. I can add an exception handler on in the `delete_chunk` method to catch these instances.

Reply:
If you use `len(xxx)` on the queryset it will load all objects into memory, it will no longer be lazy.

If you have multiple celery workers, the processing will not be linear anymore. So the deletion of Tests can start in a new job, while another job is still deleting Findings. It might not cause problems because django will still follow cascaded deletes. But I understand this PR is to avoid that and use our own hierarchy of deleting models.